addons/l10n_in_edi_ewaybill/models/account_edi_format.py

```diff
@@ -98,31 +98,30 @@ def _check_move_configuration(self, move):
                 error_message.append(_("- Transport document number and date is required when Transportation Mode is Rail,Air or Ship"))
         if error_message:
             error_message.insert(0, _("The following information are missing on the invoice (see eWayBill tab):"))
-        goods_lines = move.invoice_line_ids.filtered(lambda line: not (line.display_type in ('line_section', 'line_note', 'rounding') or line.product_id.type == "service"))
-        if not goods_lines:
+        empty_hsn_indices = []
+        empty_tax_indices = []
+        goods_lines = move.invoice_line_ids.filtered(lambda line: not (line.display_type in ('line_section', 'line_note', 'rounding'))).sorted(lambda l: (l.sequence, l.id))
+        for line_number, line in enumerate(goods_lines, start=1):
+            hsn_code = self._l10n_in_edi_extract_digits(line.l10n_in_hsn_code)
+            if not hsn_code:
+                empty_hsn_indices.append(str(line_number))
+            if not line.tax_ids:
+                empty_tax_indices.append(str(line_number))
+        if all(line.product_id.type == "service" for line in goods_lines if line.product_id):
             error_message.append(_('You need at least one product having "Product Type" as stockable or consumable.'))
+        if empty_hsn_indices:
+            error_message.append(_("Line number %(number_list)s does not have HSN/SAC code.", number_list=', '.join(empty_hsn_indices)))
+        if empty_tax_indices:
+            error_message.append(_("Line number %(number_list)s does not have tax.", number_list=', '.join(empty_tax_indices)))
         if base == "irn":
             # already checked by E-invoice (l10n_in_edi) so no need to check
             return error_message
         is_purchase = move.is_purchase_document(include_receipts=True)
         error_message += self._l10n_in_validate_partner(move.partner_id)
         error_message += self._l10n_in_validate_partner(move.company_id.partner_id, is_company=True)
         if not re.match("^.{1,16}$", is_purchase and move.ref or move.name):
-            error_message.append(_("%s number should be set and not more than 16 characters",
+            error_message.append(_("%s number cannot be more than 16 characters",
                 (is_purchase and "Bill Reference" or "Invoice")))
-        for line in goods_lines:
-            if line.product_id:
-                hsn_code = self._l10n_in_edi_extract_digits(line.l10n_in_hsn_code)
-                if not hsn_code:
-                    error_message.append(_("HSN code is not set in product %s", line.product_id.name))
-                elif not re.match("^[0-9]+$", hsn_code):
-                    error_message.append(_(
-                        "Invalid HSN Code (%s) in product %s", hsn_code, line.product_id.name
-                    ))
-            else:
-                error_message.append(_("product is required to get HSN code"))
```

Comment:
Doesn't this help the user to know what to do to have a HSN code?
This seems to have the opposite effect than the purpose of the task.

Reply:
We added HSN code in `account.move.line.` so now it's not dependent on a product we can directly add HSN code if a product is not there.